modules/rostests/apitests/acpi/Bus_PDO_QueryResourceRequirements.c

```diff
@@ -0,0 +1,524 @@
+
+typedef struct _IO_RESOURCE_REQUIREMENTS_LIST {
+  ULONG ListSize;
+  INTERFACE_TYPE InterfaceType;
+  ULONG BusNumber;
+  ULONG SlotNumber;
+  ULONG Reserved[3];
+  ULONG AlternativeLists;
+  IO_RESOURCE_LIST List[1];
+} IO_RESOURCE_REQUIREMENTS_LIST, *PIO_RESOURCE_REQUIREMENTS_LIST;
+
+/* Kernel definitions (mock) */
+#define PAGED_CODE()
+#define DPRINT1(...) do { if (0) DbgPrint(__VA_ARGS__); } while (0)
+
+typedef struct _IRP
+{
+    IO_STATUS_BLOCK IoStatus;
+} IRP, *PIRP;
+
+#define PagedPool 1
+static
+PVOID
+ExAllocatePoolWithTag(ULONG PoolType, SIZE_T NumberOfBytes, ULONG Tag)
+{
+    PVOID *Mem;
+
+    Mem = HeapAlloc(GetProcessHeap(), 0, NumberOfBytes + 2 * sizeof(PVOID));
+    Mem[0] = (PVOID)NumberOfBytes;
+    Mem[1] = (PVOID)(ULONG_PTR)Tag;
+    return Mem + 2;
+}
+
+static
+VOID
+ExFreePoolWithTag(PVOID MemPtr, ULONG Tag)
+{
+    PVOID *Mem = MemPtr;
+
+    Mem -= 2;
+    ok(Mem[1] == (PVOID)(ULONG_PTR)Tag, "Tag is %lx, expected %p\n", Tag, Mem[1]);
+    HeapFree(GetProcessHeap(), 0, Mem);
+}
+
+static
+SIZE_T
+GetPoolAllocSize(PVOID MemPtr)
+{
+    PVOID *Mem = MemPtr;
+
+    Mem -= 2;
+    return (SIZE_T)Mem[0];
+}
+
+/* ACPI driver definitions */
+typedef struct _PDO_DEVICE_DATA
+{
+    HANDLE AcpiHandle;
+    PWCHAR HardwareIDs;
+} PDO_DEVICE_DATA, *PPDO_DEVICE_DATA;
+
+/* ACPICA functions (mock) */
+static BOOLEAN AcpiCallExpected;
+static ACPI_HANDLE CorrectHandle = &CorrectHandle;
+static ACPI_BUFFER CurrentBuffer;
+static ACPI_BUFFER PossibleBuffer;
+
+ACPI_STATUS
+AcpiGetCurrentResources (
+    ACPI_HANDLE             Device,
+    ACPI_BUFFER             *RetBuffer)
+{
+    ok(AcpiCallExpected, "Unexpected call to AcpiGetCurrentResources\n");
+    ok(Device == CorrectHandle, "Device = %p, expected %p\n", Device, CorrectHandle);
+    if (RetBuffer->Length < CurrentBuffer.Length)
+    {
+        RetBuffer->Length = CurrentBuffer.Length;
+        return AE_BUFFER_OVERFLOW;
+    }
+    RetBuffer->Length = CurrentBuffer.Length;
+    CopyMemory(RetBuffer->Pointer, CurrentBuffer.Pointer, CurrentBuffer.Length);
+    return AE_OK;
+}
+
+ACPI_STATUS
+AcpiGetPossibleResources (
+    ACPI_HANDLE             Device,
+    ACPI_BUFFER             *RetBuffer)
+{
+    ok(AcpiCallExpected, "Unexpected call to AcpiGetPossibleResources\n");
+    ok(Device == CorrectHandle, "Device = %p, expected %p\n", Device, CorrectHandle);
+    if (RetBuffer->Length < PossibleBuffer.Length)
+    {
+        RetBuffer->Length = PossibleBuffer.Length;
+        return AE_BUFFER_OVERFLOW;
+    }
+    RetBuffer->Length = PossibleBuffer.Length;
+    CopyMemory(RetBuffer->Pointer, PossibleBuffer.Pointer, PossibleBuffer.Length);
+    return AE_OK;
+}
+
+#include "../../../../drivers/bus/acpi/buspdo.c"
+
+/* ACPI_RESOURCE builder helpers */
+#define MAKE_IRQ(Resource, _DescriptorLength, _Triggering, _Polarity, _Shareable, _WakeCapable) \
+    Resource->Data.Irq.DescriptorLength = _DescriptorLength;                                    \
```

Comment:
nit: safer to wrap in `do {...} while (0)`

Reply:
Yeah, will do.